### PR TITLE
Allow customizing file path for multisites and non vip sites

### LINKED ISSUE #55: Handle multisite support
Currently, it's not possible to customize the rules per network site as it's the same private directory for all of them. We should offer a way to customize this, and this could potentially be used then by those not using the plugin on VIP as well.

### PR DESCRIPTION
**Pull request by ingeniumed**
## Description

This will allow customizing the file path that's used by the plugin, for the rules file. This will be useful for multisites that want to serve a different file based on the site, as well as for non vip sites that don't have a private directory.

Currently, it includes just a site_id in the filter_options but there's plans for adding in more options down the line. This is built for extendability down the line for the filter.

This will also remove the need for Lodash, by porting the get method. That way we don't need a library workaround ever.

Fixes #55, #61 

## Steps to Test

- Use `vip dev-env` to create a multisite, with a site in it.
- In your private directory, create a folder and put a rules file there. Put another one outside it.
- Use the following in a plugin and activate it:
```php
add_filter('vip_governance__governance_file_path', __NAMESPACE__ . '\\customize_rules' , 10, 2 );

function customize_rules( $governance_file_path, $filter_options ) {
    if ( isset( $filter_options['site_id'] ) && $filter_options['site_id'] === 1 ) {
        return WPCOM_VIP_PRIVATE_DIR . '/' . WPCOMVIP_GOVERNANCE_RULES_FILENAME;
    } else if ( isset( $filter_options['site_id'] ) && $filter_options['site_id'] === 2 ) {
        return WPCOM_VIP_PRIVATE_DIR . '/' . 'site2/' . WPCOMVIP_GOVERNANCE_RULES_FILENAME;
    }

    return $governance_file_path;
}
```
- Ensure based on the site, the rules loaded is different.

Note: If you log in with the same `vipgo` user into the subsite, the rules are not going to be for a role that matches anything in the rules. It shall be for default use case only, as the `vipgo` user has no roles.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -413,6 +384,36 @@ With this rule set, the following rules will apply:
 
 There are filters in place that can be applied to change the behavior for what's allowed and what's not allowed.
 
+### `vip_governance__governance_file_path`
+
+Change the governance rules file that's used by the plugin, based on a variety of filter options that are available. By default, it is set to
+the path to `governance-rules.json` in the private directory in a VIP site. For non-vip sites, it is set to the path to `governance-rules.json` in the plugin directory.
```
Comment (smithjw1) [inline README.md]: Wouldn't it be better to say it checks first for the private directory and then to the plugin directory?

File: README.md [hunk 2]
```diff
@@ -413,6 +384,36 @@ With this rule set, the following rules will apply:
...
+
+```php
+/**
+ * Filter the governance file path, based on the filter options provided.
+ * 
+ * Currently supported keys:
+ * 
+ * site_id: The site ID for the current site.
```
Comment (smithjw1) [inline README.md:397]: I wonder what else we might include to start. post_type and author are already covered by the rules. Are there other obvious values we could include? My first thought was publish status, but I don't think that's a very good one.

If we've all thought for 5-minutes and can't come up with something obvious, we should stop trying.

**Review by smithjw1 — DISMISSED** [review]

File: README.md (hunk 1 above)
Comment (alecgeatches) [reply README.md]: Unrelated nit, was the newline between lines `389` and `390` on purpose?

File: README.md (hunk 2 above)
Comment (alecgeatches) [reply README.md:397]: I can't think of anything that the user isn't able to access themselves if they'd like. For example, even if we were missing `site_id`, user code can call into `get_current_blog_id()` directly instead, and pull other global values (like `$post`) out of the air.

It's still useful to include `site_id` and possibly other parameters for convenience, but it hopefully won't be a blocker if customer code needs to use something else.

File: governance/governance-utilities.php [hunk 3]
```diff
@@ -48,17 +48,34 @@ public static function get_governance_rules_json() {
...
+		 * 
+		 * @param string $governance_file_path Path to the governance file.
+		 * @param array $filter_options Options that can be used as a filter for determining the right file.
+		 */
+		$governance_file_path = apply_filters( 'vip_governance__governance_file_path', $governance_file_path, $filter_options );
+
 		if ( ! file_exists( $governance_file_path ) ) {
 			/* translators: %s: governance file name */
-			return new WP_Error( 'governance-file-not-found', sprintf( __( 'Governance rules (%s) could not be found in private or plugin folders.', 'vip-governance' ), WPCOMVIP_GOVERNANCE_RULES_FILENAME ) );
+			return new WP_Error( 'governance-file-not-found', sprintf( __( 'Governance rules (%s) could not be found.', 'vip-governance' ), $governance_file_path ) );
```
Comment (alecgeatches) [inline governance/governance-utilities.php]: Is there any way we can show `$governance_file_path` relative to the `wp-content` directory for the purposes of this error message? Just so we don't leak a possibly ugly internal path. So if the message looks like this:

```
Governance rules (/wp/wp-content/private/governance-rules.json) could not be found.
```

we could remove the internal paths to customer code, and return this or similar instead:

```
Governance rules (wp-content/private/governance-rules.json) could not be found.
```

If it's not too hard!

**Review by alecgeatches — DISMISSED** [review]
Changes look good to me! Left a couple of small comments, but either way good to merge.

File: README.md (hunk 1 above)
Comment (ingeniumed) [reply README.md]: Good catch, that should not be the case.

File: README.md (hunk 2 above)
Comment (ingeniumed) [reply README.md:397]: I can't think of anything else either, so would leave it as just `site_id` for now. The upside of it being a key value map is that we can add to it, without any worries for backwards compatibility.

File: governance/governance-utilities.php (hunk 3 above)
Comment (ingeniumed) [reply governance/governance-utilities.php]: That's a great point, we wouldn't want that to happen. So what I ended up doing is that, I stripped away the value of `WP_CONTENT_DIR` from the entire path. This way, what's left is going to be one of:

- Plugin path
- Private directory path
- Some other custom customer path

All of which will be within the same `/wp-content` directory. Got a screenshot of what that would look like:

Screenshot 2023-11-27 at 16 45 46

I hope I didn't miss something obvious from WP land.

**Comment by ingeniumed** [comment]
Re-requested a review from you two, as what I did new since the last review:

- Removed the need for lodash by porting a method that works for the get
- Strip the content directory from the error messages shown when a file is not found, or its contents are invalid

**Comment by alecgeatches** [comment]
@ingeniumed Noticed one thing - we can pass arbitrary paths to the filter like this:

```php
add_filter('vip_governance__governance_file_path', function( $governance_file_path, $filter_options ) {
    return '/etc/passwd';
} , 10, 2 );
```

and get file contents:

Screenshot 2023-11-27 at 9 38 16 AM

---

This isn't an massive security issue, as users need code access to modify this path anyway. However, I think it's a bit cleaner and safer if we restrict filepaths to only be allowed in `wp-content/` as well:

```php
$governance_file_path = apply_filters( 'vip_governance__governance_file_path', $governance_file_path, $filter_options );
$governance_file_path = realpath( $governance_file_path );

if ( $governance_file_path === false || substr( $governance_file_path, 0, strlen( WP_CONTENT_DIR ) ) !== WP_CONTENT_DIR ) {
	/* translators: %s: governance file name */
	return new WP_Error( 'governance-file-not-in-wp-content', sprintf( __( 'Governance rules (%s) must be in the wp-content/ directory.', 'vip-governance' ), $governance_file_path ) );
}
```

This makes it so paths like `/etc/passwd` are rejected:

image

There's room for this to be a bit cleaner (why are empty rules showing up?), but I think this is a good small improvement.

File: governance/governance-utilities.php (hunk 3 above)
Comment (alecgeatches) [reply governance/governance-utilities.php]: Perfect!

**Comment by ingeniumed** [comment]
@alecgeatches I have fixed the security problem, and it did make me realize that we really shouldn't even expose the path due to this. By not exposing it, it does make it a bit less friendly but it does clear up the security problem of accidentally revealing a path that we shouldn't. The path that's used is stil under their control, since they can provide it via code.

I have also fixed the empty rules problem as that was due to us not setting the formatted rules to be false, when there was an error. It was instead being set to an empty line, since what it was processing was an empty string. This resolves that, and makes it cleaner.

I have also, added in an extra file_Exists check because realpath can cache the path of the file and say it exists when in fact it may not. Someone pointed that out on the notes in the php docs, so I've taken that into account.

File: governance/governance-utilities.php [hunk 4]
```diff
@@ -48,17 +48,38 @@ public static function get_governance_rules_json() {
...
+		// Make sure the path is normalized. Note that file_exists() is still.
+		$governance_file_path = realpath( $governance_file_path );
+
+		// Make sure the file exists and is in the wp-content/ directory.
 		if ( ! file_exists( $governance_file_path ) ) {
-			/* translators: %s: governance file name */
-			return new WP_Error( 'governance-file-not-found', sprintf( __( 'Governance rules (%s) could not be found in private or plugin folders.', 'vip-governance' ), WPCOMVIP_GOVERNANCE_RULES_FILENAME ) );
+			return new WP_Error( 'governance-file-not-found', __( 'Governance rules could not be found.', 'vip-governance' ) );
+		} elseif ( substr( $governance_file_path, 0, strlen( WP_CONTENT_DIR ) ) !== WP_CONTENT_DIR ) {
+			return new WP_Error( 'governance-file-not-in-wp-content', __( 'Governance rules must be in the wp-content/ directory.', 'vip-governance' ) );
```
Comment (alecgeatches) [inline governance/governance-utilities.php]: Tiny suggestion here, as I'm noticing this sounds like maybe governance rules need to be directly under `wp-content/` to work. Maybe:

```suggestion
			return new WP_Error( 'governance-file-not-in-wp-content', __( 'Governance rules must be stored within the wp-content/ directory or a subdirectory.', 'vip-governance' ) );
```

**Review by alecgeatches — DISMISSED** [review]
One minor suggestion on a text change, otherwise 🚢 . Looks great!

**Review by smithjw1 — DISMISSED** [review]

**Pull request merged**